src/testController.ts

```diff
@@ -0,0 +1,207 @@
+import { exec } from "child_process";
+import { promisify } from "util";
+
+import * as vscode from "vscode";
+import { CodeLens } from "vscode-languageclient/node";
+
+import { Ruby } from "./ruby";
+import { Command } from "./status";
+
+const asyncExec = promisify(exec);
+
+export class TestController {
+  private testController: vscode.TestController;
+  private testCommands: WeakMap<vscode.TestItem, string>;
+  private testRunProfile: vscode.TestRunProfile;
+  private workingFolder: string;
+  private ruby: Ruby;
+
+  constructor(
+    context: vscode.ExtensionContext,
+    workingFolder: string,
+    ruby: Ruby
+  ) {
+    this.workingFolder = workingFolder;
+    this.ruby = ruby;
+
+    this.testController = vscode.tests.createTestController(
+      "rubyTests",
+      "Ruby Tests"
+    );
+
+    this.testCommands = new WeakMap<vscode.TestItem, string>();
+
+    this.testRunProfile = this.testController.createRunProfile(
+      "Run",
+      vscode.TestRunProfileKind.Run,
+      (request, token) => {
+        this.runHandler(request, token);
+      },
+      true
+    );
+
+    vscode.commands.executeCommand("testing.clearTestResults");
+
+    context.subscriptions.push(
+      this.testController,
+      vscode.commands.registerCommand(
+        Command.RunTest,
+        (_path, name, _command) => {
+          this.runOnClick(name);
+        }
+      )
+    );
+  }
+
+  createTestItems(response: CodeLens[]) {
+    this.testController.items.forEach((test) => {
+      this.testController.items.delete(test.id);
+    });
+
+    let classTest: vscode.TestItem;
+    const uri = vscode.Uri.from({
+      scheme: "file",
+      path: response[0].command!.arguments![0],
+    });
+
+    response.forEach((res) => {
+      if (res.data.type === "test") {
+        const [_, name, command, location] = res.command!.arguments!;
+        const testItem: vscode.TestItem = this.testController.createTestItem(
+          name,
+          name,
+          uri
+        );
+
+        this.testCommands.set(testItem, command);
+
+        testItem.range = new vscode.Range(
+          new vscode.Position(location.start_line, location.start_column),
+          new vscode.Position(location.end_line, location.end_column)
+        );
+
+        // Add test methods as children to the test class so it appears nested in Test explorer
+        // and running the test class will run all of the test methods
+        if (name.startsWith("test_")) {
+          classTest.children.add(testItem);
+        } else {
+          classTest = testItem;
+          classTest.canResolveChildren = true;
+          this.testController.items.add(testItem);
+        }
+      }
+    });
+  }
+
+  async runHandler(
+    request: vscode.TestRunRequest,
+    token: vscode.CancellationToken
+  ) {
+    const run = this.testController.createTestRun(request, undefined, true);
```

Comment:
@vinistock 

The same output terminal was being used for tests run through the Test Explorer UI, but a new terminal was created each time a test was run through code lens (i.e., through `rubyLsp.runTest`). 

Since I am using `run.appendOutput` to log the test results, I knew that `run` must contain some reference to a terminal and that terminal was the same for anything run through the Test Explorer UI but not for code lens runs. 

By debugging, I observed that every Text Explorer `TestRun` had `persisted?` set to `true` even though we were setting it to false on line 100. However, every code lens `TestRun` had `persisted?` set to `false`. This was the only difference I could find between the two runs so I set `persisted?` to true here and that fixed the issue of opening multiple terminals.